lib/api/objectPost.js

```diff
@@ -0,0 +1,137 @@
+ * @return {undefined}
+ */
+function objectPost(authInfo, request, streamingV4Params, log, callback) {
+    const {
+        headers,
+        method,
+        formData,
+        bucketName,
+    } = request;
+    let parsedContentLength = 0;
+    const passThroughStream = new PassThrough();
+    const requestType = request.apiMethods || 'objectPost';
+    const valParams = {
+        authInfo,
+        bucketName,
+        objectKey: formData.key,
+        requestType,
+        request,
+    };
+    const canonicalID = authInfo.getCanonicalID();
+
+
+    log.trace('owner canonicalID to send to data', { canonicalID });
+    return standardMetadataValidateBucketAndObj(valParams, request.actionImplicitDenies, log,
+        (err, bucket, objMD) => {
+            const responseHeaders = collectCorsHeaders(headers.origin,
+                method, bucket);
+
+            if (err && !err.AccessDenied) {
+                log.trace('error processing request', {
+                    error: err,
+                    method: 'metadataValidateBucketAndObj',
+                });
+                return callback(err, responseHeaders);
+            }
+            if (bucket.hasDeletedFlag() && canonicalID !== bucket.getOwner()) {
+                log.trace('deleted flag on bucket and request ' +
+                    'from non-owner account');
+                return callback(errors.NoSuchBucket);
+            }
+
+            return async.waterfall([
+                function countPOSTFileSize(next) {
+                    if (!request.fileEventData || !request.fileEventData.file) {
+                        return next();
+                    }
+                    request.fileEventData.file.on('data', (chunk) => {
+                        parsedContentLength += chunk.length;
+                        passThroughStream.write(chunk);
+                    });
+
+                    request.fileEventData.file.on('end', () => {
+                        passThroughStream.end();
+                        // Setting the file in the request avoids the need to make changes to createAndStoreObject's
+                        // parameters and thus all it's subsequent calls. This is necessary as the stream used to create
+                        // the object is that of the request directly; something we must work around
+                        // to use the file data produced from the multipart form data.
+                        /* eslint-disable no-param-reassign */
+                        request.fileEventData.file = passThroughStream;
+                        /* eslint-disable no-param-reassign */
+                        // Here parsedContentLength will have the total size of the file
+                        // This is used when calculating the size of the object in createAndStoreObject
+                        request.parsedContentLength = parsedContentLength;
+                        return next();
```

Comment:
That's loading the complete file in memory before sending it to sproxyd, just to get its size.

At first sight the size is used for sending to utapi before saving data. 

Once the stream is completely read, its size should be available at `bytesRead`.
Maybe if sproxyd don't need the size we could have a new code path to first stream the content to sproxyd and then get its size to send to utapi and for monitoring.

Or maybe instead we could dump it on disk on /tmp

Reply:
I now dump the file to disk then read the size with the new flow 

https://github.com/scality/cloudserver/pull/5601/files#diff-69284c3576c9c4651c9ab95928cbb676d75009ef4ee5becb238b45e22bb20efeR153

Reply:
You can't use the user provided filename as the name for the file on disk, if multiple users post the same filename at the same time they will override the data.

___

Using /tmp on disk is ok for now, but it might not be appropriate for prod as it might not scale well. /tmp is usually on the os disk and might not have enough space to scale.

I talked with the storage team, sproxyd allows sending data with chunked Transfer Encoding, so we could stream the data without knowing its Content Length beforehand.

But sproxydclient and our functions in cloudserver will need to be updated for that case. We should have another ticket for that.

Reply:
I think I need some clarity on actions to take here on my part: 

Is it to make a ticket?
If so, what are the related sproxydclient and cloudserver parts that may need changing?

Is it to make changes to the code here?
If so, should I just make a filename change for now?

Reply:
A simpler way to generate a random string `crypto.randomBytes(12).toString('hex')` you probably don't need to use the provided filename at all, but you can prefix the name with something like `cldsrv-post` so we know where the file comes from.
Or you could even use the request_id that should be in the logger.

___

For the action, yes we will need a ticket in S3C to update the streaming to sproxyd without known Content-Length and use the `Transfer-Encoding: chunked` different steps would be:
- update sproxydclient to be able to use `Transfer-Encoding: chunked`
- update cloudserver to use sproxyd with `Transfer-Encoding: chunked` and managed utapi calls afterwards
- ensure it also works on artesca with their data client